### PR TITLE
Restore git path too long

### PR DESCRIPTION
**Pull request by mingwli**
1. turn on core.longpaths on clone;
2. turn on core.longpaths before fetch;
[git-config bool case-insensitive](https://git-scm.com/docs/git-config#git-config-boolean)

### CODE REVIEW

File: src/docfx/lib/git/GitUtility.cs [hunk 1]
```diff
@@ -118,16 +119,21 @@ public static bool IsRepo(string path)
...
+                ? $"clone -c core.longpaths=true {remote} \"{path.Replace("\\", "/")}\""
+                : $"clone -c core.longpaths=true -b {branch} --single-branch {remote} \"{path.Replace("\\", "/")}\"";
 
             if (bare)
                 cmd += " --bare";
 
             return ExecuteNonQuery(cwd, cmd);
         }
 
+
```
Comment (OsmondJiang) [inline src/docfx/lib/git/GitUtility.cs]: remove two empty lines

File: src/docfx/lib/git/GitUtility.cs [hunk 2]
```diff
@@ -179,6 +185,13 @@ public static Task AddWorkTree(string cwd, string commitHash, string path)
         public static Task PruneWorkTrees(string cwd)
             => ExecuteNonQuery(cwd, $"worktree prune");
 
+        /// <summary>
+        /// Turn core.longpaths to true in current git repository.
```
Comment (OsmondJiang) [inline src/docfx/lib/git/GitUtility.cs]: the description is not matched with the method

**Comment by mingwli** [comment]
#3263

File: src/docfx/lib/git/GitUtility.cs [hunk 3]
```diff
@@ -118,9 +119,12 @@ public static bool IsRepo(string path)
         public static Task Clone(string cwd, string remote, string path, string branch = null, bool bare = false)
         {
             Directory.CreateDirectory(cwd);
+
+            // clone with configuration core.longpaths turned-on
+            // https://stackoverflow.com/questions/22575662/filename-too-long-in-git-for-windows#answer-40909460
             var cmd = string.IsNullOrEmpty(branch)
-                ? $"clone {remote} \"{path.Replace("\\", "/")}\""
-                : $"clone -b {branch} --single-branch {remote} \"{path.Replace("\\", "/")}\"";
+                ? $"clone -c core.longpaths=true {remote} \"{path.Replace("\\", "/")}\""
```
Comment (OsmondJiang) [inline src/docfx/lib/git/GitUtility.cs]: can we add a spec test for that?

File: src/docfx/restore/RestoreGit.cs [hunk 4]
```diff
@@ -93,8 +93,11 @@ async Task FetchOrCloneRepo()
             {
                 if (GitUtility.IsRepo(restoreDir))
                 {
+                    // turn on core.longpaths before fetch
+                    await GitUtility.SetCoreLongPaths(restorePath, true);
```
Comment (yufeih) [inline src/docfx/restore/RestoreGit.cs]: >SetCoreLongPaths [](start = 37, length = 16)

if the repo is cloned using `-c core.longpaths=true`, do we still need to configure it on fetch?

File: src/docfx/lib/git/GitUtility.cs (hunk 3 above)
Comment (yufeih) [reply src/docfx/lib/git/GitUtility.cs]: +1

---
In reply to: [212568157](https://github.com/dotnet/docfx/pull/3282#discussion_r212568157) [](ancestors = 212568157)

File: src/docfx/lib/git/GitUtility.cs [hunk 5]
```diff
@@ -6,6 +6,7 @@
 using System.ComponentModel;
 using System.Diagnostics;
 using System.IO;
+using System.Text;
```
[the remark this answers is not in the document]
Comment (mingwli) [reply src/docfx/lib/git/GitUtility.cs]: it's not used, will remove it.

File: src/docfx/restore/RestoreGit.cs (hunk 4 above)
Comment (mingwli) [reply src/docfx/restore/RestoreGit.cs]: If the repo is migrated from a `core.longpaths=false` state, it needs to be switched on. 
Do you recommend check `core.longpaths` before fetch and decide if it needs to be turned on?

Comment (yufeih) [reply src/docfx/restore/RestoreGit.cs]: no, I was thinking we may safely assume this repo is totally managed by *us*, so it won't have a `core.longpaths=false` state given we haven't shipped v3 yet.

Comment (OsmondJiang) [reply src/docfx/restore/RestoreGit.cs]: for migration? some repo were already cloned by us using core.longpaths=false

**Pull request merged**